be/src/util/compression/compression_context_pool.h

```diff
@@ -114,14 +114,19 @@ class CompressionContextPool {
 
 private:
     void add(InternalRef ptr) {
+        // Use explicit producer token to avoid the overhead of too many sub-ququeues
+        static thread_local std::unique_ptr<::moodycamel::ProducerToken> producer_token;
```

Comment:
this doesn't look right, the number of producer_token equals to the number of threads to be used for the compression. This depends on how the compression context is used by the caller, if the compression is performed in a fixed number of thread pool, that can be still OK, but if it is dynamic thread pool with thread create and destroy, the producer token can be still a lot.

Not sure if I understand this correct.

Reply:
You can check the implementation of `ExplicitProducer`, ProducerToken can reuse queue:
```
ProducerToken::ProducerToken(ConcurrentQueue<T, Traits>& queue)
	: producer(queue.recycle_or_create_producer(true))
{
	if (producer != nullptr) {
		producer->token = this;
	}
}
```

Reply:
```
ProducerBase* recycle_or_create_producer(bool isExplicit, bool& recycled)
	{
#ifdef MCDBGQ_NOLOCKFREE_IMPLICITPRODHASH
		debug::DebugLock lock(implicitProdMutex);
#endif
		// Try to re-use one first
		for (auto ptr = producerListTail.load(std::memory_order_acquire); ptr != nullptr; ptr = ptr->next_prod()) {
			if (ptr->inactive.load(std::memory_order_relaxed) && ptr->isExplicit == isExplicit) {
				bool expected = true;
				if (ptr->inactive.compare_exchange_strong(expected, /* desired */ false, std::memory_order_acquire, std::memory_order_relaxed)) {
					// We caught one! It's been marked as activated, the caller can have it
					recycled = true;
					return ptr;
				}
			}
		}
		
		recycled = false;
		return add_producer(isExplicit ? static_cast<ProducerBase*>(create<ExplicitProducer>(this)) : create<ImplicitProducer>(this));
	}
```

Reply:
Instead of depending on the caller's thread num, how about the context thread pool itself controls the number of total producer's token, and use thread id hash or round robin to assign to certain producer token (slot).

Reply:
Or is it overkill to use multiple-producers concurrent queue just for a memory pool usage. Is a single producer good enough to serve the usage?

Reply:
The next thing is, what if the thread lives longer than this _ctx_resources? the destruction of a thread local `producer_token` will cause invalid access to the `_ctx_resources`

Reply:
`_ctx_resources` is part of `CompressionContextPool`, and all of `CompressionContextPool` is static and global. `CompressionContextPool` will always live longer than `Ref`.